src/storage/src/hummock/file_cache/cache.rs

```diff
@@ -0,0 +1,365 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hasher;
+use std::sync::Arc;
+use std::u8;
+
+use async_trait::async_trait;
+use itertools::Itertools;
+use risingwave_common::cache::LruCache;
+use tokio::sync::Notify;
+
+use super::buffer::TwoLevelBuffer;
+use super::coding::CacheKey;
+use super::error::Result;
+use super::filter::Filter;
+use super::meta::SlotId;
+use super::store::{Store, StoreOptions, StoreRef};
+
+const INDICES_LRU_SHARD_BITS: usize = 6;
+
+pub struct FileCacheOptions {
+    pub dir: String,
+    pub capacity: usize,
+    pub total_buffer_capacity: usize,
+    pub cache_file_fallocate_unit: usize,
+    pub filters: Vec<Arc<dyn Filter>>,
+
+    pub flush_buffer_hooks: Vec<Arc<dyn FlushBufferHook>>,
+}
+
+#[async_trait]
+pub trait FlushBufferHook: Send + Sync + 'static {
+    async fn pre_flush(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn post_flush(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+struct BufferFlusher<K>
+where
+    K: CacheKey,
+{
+    buffer: TwoLevelBuffer<K>,
+    store: StoreRef<K>,
+    indices: Arc<LruCache<K, SlotId>>,
+    notifier: Arc<Notify>,
+
+    hooks: Vec<Arc<dyn FlushBufferHook>>,
+}
+
+impl<K> BufferFlusher<K>
+where
+    K: CacheKey,
+{
+    async fn run(&self) -> Result<()> {
```

Comment:
The current TwoLevelBuffer design only works with one BufferFlusher thread. I am not sure what the expected write throughput for the cache but I suspect this cannot saturate the disk, especially for SSD.

Reply:
The write model is designed for cloud block device like EBS only. Punch holes might not be adaptable to SSDs, either, because that may cause SSD GC more frequently when there is not inner free blocks. 🤔